caddy_scraper/opensearch_query_engine.py

```diff
@@ -0,0 +1,144 @@
+from opensearchpy import OpenSearch
+import requests
+from langchain_community.embeddings import BedrockEmbeddings
+from langchain.schema import Document
+from typing import List
+
+
+class OpenSearchQueryEngine:
+    def __init__(
+        self,
+        client: OpenSearch,
+        index_name: str = "caddy-hybrid-search-index",
+        lexical_weight: float = 0.3,
+        vector_weight: float = 0.7
+    ):
+        self.client = client
+        self.index_name = index_name
+        self.embedding_model = BedrockEmbeddings(
+            model_id="cohere.embed-english-v3", region_name="eu-west-3"
+        )
+        # Create the search pipeline for normalization
+        self._create_search_pipeline(lexical_weight, vector_weight)
+
+    def _create_search_pipeline(self, lexical_weight: float, vector_weight: float):
+        """Create a search pipeline with normalization processor"""
+        pipeline_config = {
+            "description": "Post processor for hybrid search",
+            "phase_results_processors": [{
+                "normalization-processor": {
+                    "normalization": {
+                        "technique": "min_max"
+                    },
+                    "combination": {
+                        "technique": "arithmetic_mean",
+                        "parameters": {
+                            "weights": [lexical_weight, vector_weight]  
+                        }
+                    }
+                }
+            }]
+        }
+        
+        try:
+            self.client.transport.perform_request(
```

Comment:
We're currently configuring each time we construct an engine, but this preliminary setup for the OpenSearch instance?

Reply:
Definitely not wedded to where this happens. In this PR I tried to make a distinction between upload (document-management) and retrieval (query_engine) and this felt more closely related to the later. Only benefit to keeping it in a query_engine is the ability to tweak hybrid search weightings at retrieval time. I agree that once we have stopped experimenting it makes sense to move this to a separate configuration process